### PR TITLE
Create script to fix birthday, anonymizing day and month.

### PR DESCRIPTION
**Pull request by ashercodeorg**
Note that the script is intentionally idempotent.

**Note that this script mutates the age for older oauth accounts. In particular, any such user that is twenty-one+ on 2017-05-31 is now exactly twenty-one. Thus, we can no longer distinguish between a twenty-one+ year old from a twenty-three+ year old.**

TESTING:
- [x] Ran the script locally, with a mix of data, confirming expected results.

UPDATE:
- [x] Run on staging.
- [x] Run on levelbuilder.
- [x] Run on test.
- [x] Run on production.

### CODE REVIEW

File: bin/oneoff/data_fix/birthday_and_created_at [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# DAY(birthday) == DAY(created_at), adjusting YEAR(birthday) as necessary, to conform to our privacy
+# policy of storing ages (not birthdays).
+
+require_relative '../../../dashboard/config/environment'
+
+updated_count = 0
+User.with_deleted.find_each do |user|
+  # Perform some console logging to give some user feedback on progress.
+  if updated_count % 1_000 == 0
+    puts "UPDATED_COUNT: #{updated_count}"
```
Comment (ashercodeorg) [inline bin/oneoff/data_fix/birthday_and_created_at]: Note that this may output multiple times, as subsequent users are `next`ed. This seems fine to me.

**Comment by ashercodeorg** [comment]
Ping.

File: bin/oneoff/data_fix/birthday_and_created_at [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  # better as the result of half of our accounts needing fixing. Consequently, it seems better to
+  # iterate over all accounts rather than paging through (a random collection) half the accounts.
+  next unless user[:birthday]
+  if user[:birthday].month == user[:created_at].month &&
+    user[:birthday].day == user[:created_at].day
+    next
+  end
+
+  # Though not the cleanest, this logic mirrors the logic in User#from_omniauth.
+  user.birthday = auth.info.dob
```
Comment (jeremydstone) [inline bin/oneoff/data_fix/birthday_and_created_at]: Where does `auth` come from?

**Comment by jeremydstone** [comment]
What testing have you done on this so far? Would it make sense to run in production in a limited fashion first? In the unlikely case there is some bug, we'd like to figure that out before we ran this on all of our users. Would it make sense to do something like run against 1,000 or 10,000 users first and check the results? Or pick out a known good case and known bad case and run against both of them individually to check?

File: bin/oneoff/data_fix/birthday_and_created_at (hunk 2 above)
Comment (ashercodeorg) [reply bin/oneoff/data_fix/birthday_and_created_at]: Oops, that line needs to be removed.

**Comment by ashercodeorg** [comment]
No testing thus far. EDIT: Tested locally, on users that did and did not need an update.

After it gets reviewed, I'll run on `test`, then `staging` (which has hundreds of rows needing fixing).

**Comment by ashercodeorg** [comment]
PTAL.

**Comment by jeremydstone** [comment]
LGTM. When running in production, you might consider a crawl/walk/run strategy. On things like this, I will frequently start with a small batch (as small as 1) and verify in production that the right thing happened, then increase to some bigger batches before ultimately letting it run everything.

**Pull request merged**